controllers/spacerequest/spacerequest_controller.go

```diff
@@ -399,6 +412,125 @@ func (r *Reconciler) deleteExistingSubSpace(logger logr.Logger, subSpace *toolch
+		}
+		if err := memberClusterWithSpaceRequest.Client.List(context.TODO(), secretList, secretLabels, client.InNamespace(spaceRequest.GetNamespace())); err != nil {
+			return errs.Wrap(err, fmt.Sprintf(`attempt to list Secrets associated with spaceRequest %s in namespace %s failed`, spaceRequest.GetName(), spaceRequest.GetNamespace()))
+		}
+
+		kubeConfigSecret := &v1.Secret{}
+		switch {
+		case len(secretList.Items) == 0:
+			// create the secret for this namespace
+			clientConfig, err := r.generateKubeConfig(memberClusterWithSpaceRequest, namespace.Name)
+			if err != nil {
+				return err
+			}
+			clientConfigFormatted, err := clientcmd.Write(*clientConfig)
+			if err != nil {
+				return err
+			}
+			kubeConfigSecret = &v1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: spaceRequest.Name + "-",
+					Namespace:    spaceRequest.Namespace,
+					Labels:       secretLabels,
+				},
+				Type: v1.SecretTypeOpaque,
+				StringData: map[string]string{
+					"kubeconfig": string(clientConfigFormatted),
+				},
+			}
+			if err := controllerutil.SetControllerReference(spaceRequest, kubeConfigSecret, r.Scheme); err != nil {
+				return errs.Wrap(err, "error setting controller reference for secret "+kubeConfigSecret.Name)
+			}
+			if err := memberClusterWithSpaceRequest.Client.Create(context.TODO(), kubeConfigSecret); err != nil {
+				return err
+			}
+			logger.Info("Created Secret", "Name", kubeConfigSecret.Name)
+
+			// add provisioned namespace and name of the secret that provides access to the namespace
+			namespaceAccess = append(namespaceAccess, toolchainv1alpha1.NamespaceAccess{
+				Name:      namespace.Name,
+				SecretRef: kubeConfigSecret.Name,
+			})
+
+		case len(secretList.Items) == 1:
+			// a secret is already present for this namespace
+			namespaceAccess = append(namespaceAccess, toolchainv1alpha1.NamespaceAccess{
+				Name:      namespace.Name,
+				SecretRef: secretList.Items[0].Name,
+			})
+
+		case len(secretList.Items) > 1:
+			// some unexpected issue causing to many secrets
+			return fmt.Errorf("invalid number of secrets found. actual %d, expected %d", len(secretList.Items), 1)
+		}
+
+	}
+
+	// update space request status in case secrets for provisioned namespace were created.
+	if len(namespaceAccess) > 0 {
+		spaceRequest.Status.NamespaceAccess = namespaceAccess
+		return memberClusterWithSpaceRequest.Client.Status().Update(context.TODO(), spaceRequest)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) generateKubeConfig(memberClusterWithSpaceRequest cluster.Cluster, namespace string) (*api.Config, error) {
+	// create a token request for the admin service account
+	token, err := restclient.CreateTokenRequest(memberClusterWithSpaceRequest.RESTClient, types.NamespacedName{
+		Namespace: namespace,
+		Name:      toolchainv1alpha1.AdminServiceAccountName,
+	}, TokenRequestExpirationSeconds)
+	if token == "" || err != nil {
+		return nil, err
+	}
```

Comment:
Actually I had hard time trying to force the client-go library to return a `nil` response. Unless I'm missing something, this doesn't seem to be possible, even when the http request returns `nil` the client go decoder still returns the `TokenRequest` object we passed to the `.Into(result)` function , which results "only" in an empty `TokenRequest.Status.Token`.

I've also moved the empty token check to `toolchain-common` in https://github.com/codeready-toolchain/host-operator/pull/789/commits/d63629163306a867e7358fa11e647a32577925c1

Reply:
If this was about `Do(context.TODO()).Into(result)`, I had the same concern but checked it yesterday and it looks like if there is a problem getting the token then there will be an error so the token shouldn't be nil.

See https://github.com/kubernetes/client-go/blob/2666bd29867c96168c5e9429fd74fd9bfbedac8b/rest/request.go#L925C16-L928

Reply:
My concert was regarding this if-statement:
```go
       if token == "" || err != nil {
		return nil, err
	}
```
from that if statement it looks like that it may happen when
```
token == ""
```
and at the same time
```
err == nil
```
which would mean that we would return:
```
return nil, nil
```


so I think that we should either change the statement to:
```go
       if err != nil {
		return nil, err
	}
```
and don't care about the content of the token, because we always expect that it should be returned


or change it to:
```go
        if err != nil {
		return nil, err
	}
        if token == "" {
		return nil, fmt.Errorf("Token is empty")
	}
```
I'm totally fine with either way

Reply:
Yeah, as @rajivnathan said, I've observed the same and moved the check:
```
        if err != nil {
		return "", err
	}
        if len(result.Status.Token) == 0 {
		return "", fmt.Errorf("unable to create token, got empty string")
	}
```

 to toolchain-common pr https://github.com/codeready-toolchain/toolchain-common/pull/310
 
 Thanks both for checking !